PyAlbany/python/RandomizedCompression.py

```diff
@@ -0,0 +1,342 @@
+# ***********************************************************************
+# @HEADER
+
+# This file contains routines for randomized truncated singular value decompositions
+# of a matrix free operator (Op), by singlePass and doublePass algorithms.
+# A routine is available for the hierarchical off-diagonal low-rank (HODLR) decomposition
+# of a matrix free operator (Op). 
+#
+# See "Finding structure with randomness, probabilistic algorithms for constructing approximate
+# matrix decompositions", N. Halko, P.G. Martinsson, J.A. Tropp SIAM Review (2011),
+# "Compressing rank-structured matrices via randomized sampling", P.G. Martinsoon, SIAM Journal
+# on Scientific Computing (2016) and "Hierarchical off-diagonal low-rank approximation of
+# Hessians in inverse problems", T. Hartland, G. Stadler, M. Perego, K. Liegeois, N. Petra (to be submitted).  
+#
+# Various helper functions are utilized to allow for a parallel implementation of the HODLR compression
+# algorithm, such as Hpartition, MPIpartition and HidxMPIproc
+#
+#
+#
+# Last updated on April 19th, 2022
+# Author Tucker Hartland, University of California, Merced
+
+
+
+from PyTrilinos import Tpetra
+from PyTrilinos import Teuchos
+try:
+    from PyAlbany import wpyalbany as wpa
+except:
+    import wpyalbany as wpa
+
+try:
+    from PyAlbany import Utils as utils
+except:
+    import Utils as utils
+
+import numpy as np
+import scipy.linalg as spla
+
+
+"""
+  singlePass: randomized singular value decomposition of a symmetric operator
+  input: Op, operator which contains a Map member data and a 
+                  dot method which takes a Tpetra MultiVector
+                             and returns a Tpetra MultiVector 
+          k, the rank of the output operator U \Lambda U^T
+  output: \Lambda, U
+          \Lambda, eigenvalues,  a nondistributed numpy array with k entries
+          U,       eigenvectors, a distributed Tpetra MultiVector with k columns
+"""
+def singlePass(Op, k):
+    comm       = Teuchos.DefaultComm.getComm()
+    rank       = comm.getRank()
+    nprocs     = comm.getSize()
+    nElems     = Op.Map.getNodeNumElements()
+    N          = Op.Map.getGlobalNumElements()
+    nMaxOrthog = min(50, k)
+
+    omega = Tpetra.MultiVector(Op.Map, k, dtype="d")
+    q     = Tpetra.MultiVector(Op.Map, k, dtype="d")
+    u     = Tpetra.MultiVector(Op.Map, k, dtype="d")
+    
+    for i in range(k):
+        omega[i, :] = np.random.randn(nElems)
+    y = Op.dot(omega)
```

Comment:
I like the notion of Operator, I did something similar in ExtremeEvent.py for `class HessianOperator(slinalg.LinearOperator)`.

I think that it would be best to gather the common operator class into a same python file to avoid having the same (or similar) thing in different files.
I am not sure if you have to do this in this PR though.
I can do it later.

Reply:
I see what you mean about having a common way to define this HessianOperator class. One thing is that I wanted the class to have a method to take in a `MultiVector` and output a `MultiVector`, but I suppose this could be accomodated by writing a `matmat` method. Let me know if you would like me to make these changes.